Add developer metrics calculation routine

seemosss opened the pull request.

Commit:
Add developer metrics calculation routine

Commit:
Add developer metrics calculation routine

crystaldust commented on dags/oss_know/libs/metrics/developer_metrics.py:
```diff
@@ -0,0 +1,132 @@
+                                  "review_dismissed", "transferred",
+                                  "unlocked", "unpinned", "user_blocked"]
+
+        privilege_sql_ = f"SELECT * FROM github_issues_timeline WHERE search_key__repo = '{self.repo}'"
```
- 查询时候的限定条件应该增加search_key__owner='{self.owner}'，数据表设计里owner对应github的organization，repo对应代码仓的名称，如[https://github.com/rust-lang/rust](https://github.com/rust-lang/rust)项目，对应owner为`rust-lang`，repo为`rust`
- 可以只select需要的字段(event_type, timeline_raw)，加快查询速度。同时在后续的event处理中，利用python的语法糖，代码编写也会比较方便，如：
```python
privilege_sql = f"SELECT search_key__event, timeline_raw FROM github_issues_timeline WHERE search_key__repo = '{self.repo}'"
privilege_results = self.clickhouse_client.execute_no_params(privilege_sql_)
for (event_type, timeline_raw) in privilege_results:
    # analyze the data
```

- SQL中的where条件可以直接限定event_type（即search_key__event列），利用clickhouse分布式处理的能力，加快处理速度。可以参考[clickhouse array的has方法](https://clickhouse.com/docs/en/sql-reference/functions/array-functions#hasarr-elem)来过滤privilege event，sample：
```sql
with ['added_to_project', 'converted_note_to_issue',
    'deployed', 'deployment_environment_changed',
    'locked', 'merged', 'moved_columns_in_project',
    'pinned', 'removed_from_project',
    'review_dismissed', 'transferred',
    'unlocked', 'unpinned', 'user_blocked'] as privileged_events
select search_key__event, timeline_raw
from github_issues_timeline
where search_key__owner = 'rust-lang'
  and search_key__repo = 'rust'
  and has(privileged_events, search_key__event)
limit 10
```

crystaldust commented on dags/oss_know/libs/metrics/developer_metrics.py:
```diff
@@ -0,0 +1,132 @@
+                dev_name = dev['login']
+                if dev_name not in privilege_map:
+                    privilege_map[dev_name] = [0] * 14
+                index = privileged_events_list.index(event)
```
这里会报错提示privileged_events_list index out of range。根据上下文，这里传入的参数应该是event_type而非event？

可以尝试利用python的dict类型来存储一份event_type到index的对应关系：
```python
event_type_index_map = {}
for (index, event) in enumerate(privileged_events_list):
    event_type_index_map[event] = index
```
这样，在获取index时，只需访问`event_type_index_map[event_type]`就可以获得对应的index。一般来说，使用哈希表查找速度比list要快。

crystaldust commented on dags/oss_know/libs/metrics/developer_metrics.py:
```diff
@@ -0,0 +1,132 @@
+
+class CountMetricRoutineCalculation(MetricRoutineCalculation):
+    def calculate_metrics(self):
+        gits_sql_ = f"SELECT * FROM gits WHERE search_key__repo = '{self.repo}'"
```
增加search_key__owner的限定条件，只获取必需的字段

crystaldust commented on dags/oss_know/libs/metrics/developer_metrics.py:
```diff
@@ -0,0 +1,132 @@
+
+class NetworkMetricRoutineCalculation(MetricRoutineCalculation):
+    def calculate_metrics(self):
+        gits_sql_ = f"SELECT * FROM gits WHERE search_key__repo = '{self.repo}'"
```
增加search_key__owner的限定条件，只获取必需的字段

crystaldust commented on dags/oss_know/libs/metrics/developer_metrics.py:
```diff
@@ -0,0 +1,132 @@
+            degree = social_network.degree(dev)
+            response.append((dev, degree, eigenvector[dev]))
+
+        logger.info('calculating  Network Metrics')
```
日志里可以加上当前在处理的owner和repo信息，以及其他信息，如graph的edge、node数量等，这样在查询日志时可以看到一些关键信息。

crystaldust commented on dags/oss_know/libs/metrics/developer_metrics.py:
```diff
@@ -0,0 +1,132 @@
+
+class PrivilegeEventsMetricRoutineCalculation(MetricRoutineCalculation):
+    def calculate_metrics(self):
+        privileged_events_list = ["added_to_project", "converted_note_to_issue",
```
这里的privileged events都是固定的字段，可以存成类的static成员，不必每次计算都声明一次变量

Commit:
modify developer metrics

Commit:
modify developer metrics

crystaldust committed:
Update ref to influence metrics repo vars and routine interval.

crystaldust committed:
Optimize iteration on routine classes and variables names

crystaldust committed:
Add deps on networkx and mysql connector in Dockerfile and requirments files

crystaldust committed:
[TEMP]
Multiple optimizations and updates on developer metrics calculation

- logs
- sql optimizations
- privileged events calculation not seems to be expected, discussion
  needed

crystaldust committed:
Fix the privileged events row construction

crystaldust committed:
Optimize developer contrib count metric calculation.

Calculate commit count and total lines of code in SQL instead of
manually accumulation for better performance.

crystaldust committed:
up

crystaldust committed:
Calculate co-modified files by SQL.

Co-authored-by: Lin Ma<malin3598@foxmail.com>

crystaldust committed:
Use a different table name and add owner, repo cols.

The table network_metrics conflicts with other metrics.

seemosss committed:
Merge branch 'development' into zhourun-dev

seemosss committed:
Merge pull request #1 from crystaldust/zhourun-dev

Improve privileged events and developer network metric calculations

crystaldust commented:
Close the PR which is obtained in #213
